File: backend/app/services/ballistics.py
import logging
from typing import List
from ..models import (
    CalculationRequest,
    CalculationResponse,
    TrajectoryPoint,
    DragModelEnum
)

# Import py-ballisticcalc components
from py_ballisticcalc import (
    Calculator, DragModel, Ammo, Weapon, Atmo, Shot, Wind,
    TableG1, TableG7
)
from py_ballisticcalc.unit import (
    Distance, Velocity, Temperature, Weight, Pressure, Angular, Energy
)

logger = logging.getLogger(__name__)


class BallisticsService:
    """Service for performing ballistics calculations."""

    def __init__(self):
        self.calculator = Calculator()

    async def calculate_trajectory(
        self,
        request: CalculationRequest
    ) -> CalculationResponse:
        """Calculate trajectory based on input parameters."""
        try:
            logger.info(
                f"Starting trajectory calculation for range "
                f"{request.max_range} yards"
            )

            # Create drag model
            drag_table = self._get_drag_table(request.ammo.drag_model)
            drag_model = DragModel(request.ammo.bc, drag_table)

            # Create ammo object
            ammo = Ammo(
                drag_model,
                Velocity.FPS(request.ammo.muzzle_velocity),
                Temperature.Fahrenheit(request.atmosphere.temperature)
            )

            # Add bullet weight if provided
            if request.ammo.bullet_weight:
                ammo.bullet_weight = Weight.Grain(request.ammo.bullet_weight)

            # Create weapon object
            weapon = Weapon(
                sight_height=Distance.Inch(request.weapon.sight_height)
            )
            if request.weapon.twist:
                weapon.twist = Distance.Inch(request.weapon.twist)

            # Create atmospheric conditions
            atmo = Atmo(
                altitude=Distance.Foot(request.atmosphere.altitude),
                temperature=Temperature.Fahrenheit(
                    request.atmosphere.temperature
                ),
                pressure=Pressure.InHg(request.atmosphere.pressure),
                humidity=request.atmosphere.humidity
            )

            # Create shot with wind
            shot = Shot(weapon=weapon, ammo=ammo, atmo=atmo)
            if request.wind.speed > 0:
                shot.winds = [Wind(
                    Velocity.MPH(request.wind.speed),
                    Angular.OClock(request.wind.direction)
                )]

            # Set zero and calculate trajectory
            zero_distance = Distance.Yard(request.zero_distance)
            logger.info(f"Zero distance: {request.zero_distance} yds")
            zero_adjustment = self.calculator.set_weapon_zero(
                shot, zero_distance
            )
            logger.info(f"Zero adjustment calculated: {zero_adjustment}")

            # Calculate trajectory with step size
            trajectory_range = Distance.Yard(request.max_range)
            trajectory_step = Distance.Yard(request.step_size)
            max_range = request.max_range
            step_size = request.step_size
            logger.info(f"Range: {max_range} yds, Step: {step_size} yds")

            result = self.calculator.fire(
                shot,
                trajectory_range=trajectory_range,
                trajectory_step=trajectory_step,
                extra_data=True
            )

            # Extract trajectory data from HitResult object
            # py-ballisticcalc returns a HitResult containing a trajectory list
            if hasattr(result, 'trajectory'):
                all_trajectory_points = result.trajectory
            else:
                # Fallback for unexpected result structure
                all_trajectory_points = [result]

            # Filter trajectory points to only include clean step intervals
            # py-ballisticcalc includes extra points for zeroing calculations
            # and internal ballistics computations that we don't want to show
            trajectory_data = []
            for point in all_trajectory_points:
                distance = float(point.distance >> Distance.Yard)
                # Include muzzle point (0) and points at exact step intervals
                if distance == 0.0 or distance % step_size == 0.0:
                    trajectory_data.append(point)

            logger.info(
                f"Filtered to {len(trajectory_data)} points from "
                f"{len(all_trajectory_points)} total points"
            )

            # Convert to response format
            trajectory_points = self._convert_trajectory_points(
                trajectory_data, request
            )

            logger.info(
                f"Trajectory calculation completed with "
                f"{len(trajectory_points)} points"
            )

            return CalculationResponse(
                trajectory=trajectory_points,
                zero_adjustment=float(zero_adjustment >> Angular.Mil),
                success=True,
                message="Calculation completed successfully"
            )

        except Exception as e:
            logger.error(f"Trajectory calculation failed: {str(e)}")
            raise ValueError(f"Calculation error: {str(e)}")

    def _get_drag_table(self, drag_model: DragModelEnum):
        """Get the appropriate drag table."""
        if drag_model == DragModelEnum.G1:
            return TableG1
        elif drag_model == DragModelEnum.G7:
            return TableG7
        else:
            return TableG1

    def _convert_trajectory_points(
        self, result, request: CalculationRequest
    ) -> List[TrajectoryPoint]:
        """Convert calculation result to trajectory points."""
        trajectory_points = []

        for point in result:
            # Extract energy value from py-ballisticcalc or calculate manually
            energy = 0.0
            if hasattr(point, 'energy') and point.energy is not None:
                try:
                    energy = float(point.energy >> Energy.FootPound)
                except Exception:
                    energy = 0.0

            # If py-ballisticcalc energy is 0, calculate manually
            if energy == 0.0 and request.ammo.bullet_weight:
                # Manual kinetic energy calculation: KE = 0.5 * m * v²
                # Convert grains to pounds (7000 grains = 1 pound)
                mass_pounds = request.ammo.bullet_weight / 7000.0
                velocity_fps = float(point.velocity >> Velocity.FPS)
                # Energy in foot-pounds (divide by gc = 32.174 lbm⋅ft/(lbf⋅s²))
                energy = 0.5 * mass_pounds * (velocity_fps ** 2) / 32.174

            # Extract drop adjustment in Mil
            drop_adj = 0.0
            if hasattr(point, 'drop_adj'):
                drop_adj = float(point.drop_adj >> Angular.Mil)
            elif hasattr(point, 'drop_adjustment'):
                drop_adj = float(point.drop_adjustment >> Angular.Mil)

            # Extract windage adjustment in Mil
            windage_adj = 0.0
            if hasattr(point, 'windage_adj'):
                windage_adj = float(point.windage_adj >> Angular.Mil)
            elif hasattr(point, 'windage_adjustment'):
                windage_adj = float(point.windage_adjustment >> Angular.Mil)

            # Try different attribute names for drop/height
            drop_value = 0.0
            if hasattr(point, 'height'):
                drop_value = float(point.height >> Distance.Inch)
            elif hasattr(point, 'slant_height'):
                drop_value = float(point.slant_height >> Distance.Inch)
            elif hasattr(point, 'y'):
                drop_value = float(point.y >> Distance.Inch)
            elif hasattr(point, 'drop'):
                drop_value = float(point.drop >> Distance.Inch)

            # Try different attribute names for windage
            windage_value = 0.0
            if hasattr(point, 'windage'):
                windage_value = float(point.windage >> Distance.Inch)
            elif hasattr(point, 'z'):
                windage_value = float(point.z >> Distance.Inch)

            trajectory_points.append(TrajectoryPoint(
                distance=float(point.distance >> Distance.Yard),
                drop=drop_value,
                windage=windage_value,
                velocity=float(point.velocity >> Velocity.FPS),
                energy=energy,
                time=float(point.time),
                drop_adjustment=drop_adj,
                windage_adjustment=windage_adj
            ))

        return trajectory_points

    def get_available_drag_models(self) -> List[str]:
        """Get list of available drag models."""
        return [model.value for model in DragModelEnum]

File: frontend/src/lib/ballistics.ts
import {
  Ammo,
  Angular,
  Atmo,
  Calculator as JsBallisticsCalculator,
  DragModel,
  Energy,
  Shot,
  Table,
  TrajectoryData,
  Velocity,
  Weapon,
  Wind,
  Distance,
  UNew,
  calculateEnergy,
} from "js-ballistics";

export const LIMITS = {
  MAX_RANGE_YARDS: 3000,
  MIN_RANGE_YARDS: 25,
  MAX_STEP_SIZE: 100,
  MIN_STEP_SIZE: 1,
} as const;

const DRAG_TABLES = {
  G1: Table.G1,
  G7: Table.G7,
} as const;

export type DragModelName = keyof typeof DRAG_TABLES;

export interface WeaponInput {
  sight_height: number;
  twist?: number | null;
}

export interface AmmoInput {
  bc: number;
  drag_model: DragModelName;
  muzzle_velocity: number;
  bullet_weight?: number | null;
  bullet_diameter?: number | null;
}

export interface AtmosphereInput {
  temperature: number;
  pressure: number;
  humidity: number;
  altitude: number;
}

export interface WindInput {
  speed: number;
  direction: number;
}

export interface CalculationRequest {
  weapon: WeaponInput;
  ammo: AmmoInput;
  atmosphere: AtmosphereInput;
  wind: WindInput;
  zero_distance: number;
  max_range: number;
  step_size: number;
}

export interface TrajectoryPoint {
  distance: number;
  drop: number;
  windage: number;
  velocity: number;
  energy: number;
  time: number;
  drop_adjustment: number;
  windage_adjustment: number;
}

export interface CalculationResponse {
  trajectory: TrajectoryPoint[];
  zero_adjustment: number;
  success: boolean;
  message: string;
}

export interface ValidationResult {
  valid: boolean;
  message: string;
  estimated_points: number;
}

type CalculatorConfig = Record<string, never>;
type BulletWeight = number | null;

const STEP_TOLERANCE = 1e-4;

export class BallisticsError extends Error {
  constructor(message: string, public readonly code: string = "ballistics_error") {
    super(message);
    this.name = "BallisticsError";
  }
}

export class BallisticsCalculator {
  private readonly calculator: JsBallisticsCalculator<CalculatorConfig>;

  constructor(calculator?: JsBallisticsCalculator<CalculatorConfig>) {
    this.calculator = calculator ?? new JsBallisticsCalculator<CalculatorConfig>();
  }

  getAvailableDragModels(): DragModelName[] {
    return Object.keys(DRAG_TABLES) as DragModelName[];
  }

  validate(request: CalculationRequest): ValidationResult {
    if (request.max_range > LIMITS.MAX_RANGE_YARDS) {
      return {
        valid: false,
        message: `Maximum range cannot exceed ${LIMITS.MAX_RANGE_YARDS} yards`,
        estimated_points: 0,
      };
    }

    if (request.step_size > LIMITS.MAX_STEP_SIZE) {
      return {
        valid: false,
        message: `Step size cannot exceed ${LIMITS.MAX_STEP_SIZE} yards`,
        estimated_points: 0,
      };
    }

    if (request.step_size < LIMITS.MIN_STEP_SIZE) {
      return {
        valid: false,
        message: `Step size must be at least ${LIMITS.MIN_STEP_SIZE} yard`,
        estimated_points: 0,
      };
    }

    if (request.zero_distance < LIMITS.MIN_RANGE_YARDS) {
      return {
        valid: false,
        message: `Zero distance must be at least ${LIMITS.MIN_RANGE_YARDS} yards`,
        estimated_points: 0,
      };
    }

    if (request.max_range <= request.zero_distance) {
      return {
        valid: false,
        message: "Maximum range must be greater than zero distance",
        estimated_points: 0,
      };
    }

    const estimated_points = Math.floor((request.max_range - request.zero_distance) / request.step_size) + 1;

    return {
      valid: true,
      message: "Parameters are valid",
      estimated_points,
    };
  }

  calculate(request: CalculationRequest): CalculationResponse {
    const validation = this.validate(request);
    if (!validation.valid) {
      throw new BallisticsError(validation.message, "validation_error");
    }

    const { shot, bulletWeightGrains } = this.buildShot(request);

    const zeroDistance = UNew.Yard(request.zero_distance);
    const zeroAngle = this.calculator.setWeaponZero(shot, zeroDistance);

    const hitResult = this.calculator.fire({
      shot,
      trajectoryRange: UNew.Yard(request.max_range),
      trajectoryStep: UNew.Yard(request.step_size),
      extraData: true,
    });

    const filtered = this.filterTrajectory(hitResult.trajectory, request.step_size);
    const trajectory = filtered.map((point) =>
      this.toTrajectoryPoint(point, bulletWeightGrains)
    );

    return {
      trajectory,
      zero_adjustment: zeroAngle.In(Angular.MIL),
      success: true,
      message: "Calculation completed successfully",
    };
  }

  private buildShot(request: CalculationRequest) {
    const dragTable = DRAG_TABLES[request.ammo.drag_model] ?? DRAG_TABLES.G1;

    const bulletWeightGrains = request.ammo.bullet_weight ?? null;
    const dragModel = new DragModel({
      bc: request.ammo.bc,
      dragTable,
      weight: bulletWeightGrains ? UNew.Grain(bulletWeightGrains) : 0,
      diameter: request.ammo.bullet_diameter ? UNew.Inch(request.ammo.bullet_diameter) : 0,
    });

    const ammo = new Ammo({
      dm: dragModel,
      mv: UNew.FPS(request.ammo.muzzle_velocity),
      powderTemp: UNew.Fahrenheit(request.atmosphere.temperature),
    });

    const weapon = new Weapon({
      sightHeight: UNew.Inch(request.weapon.sight_height),
      twist: request.weapon.twist ? UNew.Inch(request.weapon.twist) : null,
    });

    const atmo = new Atmo({
      altitude: UNew.Foot(request.atmosphere.altitude),
      pressure: UNew.InHg(request.atmosphere.pressure),
      temperature: UNew.Fahrenheit(request.atmosphere.temperature),
      humidity: request.atmosphere.humidity,
      powderT: UNew.Fahrenheit(request.atmosphere.temperature),
    });

    const winds = request.wind.speed > 0
      ? [
          new Wind({
            velocity: UNew.MPH(request.wind.speed),
            directionFrom: UNew.OClock(request.wind.direction),
          }),
        ]
      : undefined;

    const shot = new Shot({ weapon, ammo, atmo, winds });

    return { shot, bulletWeightGrains };
  }

  private filterTrajectory(points: TrajectoryData[], stepSize: number): TrajectoryData[] {
    return points.filter((point) => {
      const distanceYards = point.distance.In(Distance.Yard);
      if (Math.abs(distanceYards) < STEP_TOLERANCE) {
        return true;
      }

      const ratio = distanceYards / stepSize;
      return Math.abs(ratio - Math.round(ratio)) < STEP_TOLERANCE;
    });
  }

  private toTrajectoryPoint(point: TrajectoryData, bulletWeightGrains: BulletWeight): TrajectoryPoint {
    const distance = point.distance.In(Distance.Yard);
    const velocity = point.velocity.In(Velocity.FPS);

    const dropCandidates = [
      point.height?.In(Distance.Inch),
      point.targetDrop?.In(Distance.Inch),
    ];
    const drop = dropCandidates.find((value) => Number.isFinite(value)) ?? 0;

    const windage = point.windage?.In(Distance.Inch) ?? 0;
    let energy = point.energy?.In(Energy.FootPound) ?? 0;

    if ((!Number.isFinite(energy) || energy === 0) && bulletWeightGrains) {
      energy = calculateEnergy(bulletWeightGrains, velocity);
    }

    return {
      distance,
      drop,
      windage,
      velocity,
      energy,
      time: point.time,
      drop_adjustment: point.dropAdjustment?.In(Angular.MIL) ?? 0,
      windage_adjustment: point.windageAdjustment?.In(Angular.MIL) ?? 0,
    };
  }
}

export const ballisticsCalculator = new BallisticsCalculator();
